stonesoup/smoother/kalman.py

```diff
@@ -0,0 +1,288 @@
+
+        Parameters
+        ----------
+        state : :class:`~.State`
+            The input state (to check for a linked prediction)
+        **kwargs
+            These are passed to the :meth:`matrix()` function
+
+        Returns
+        -------
+         : :class:`numpy.ndarray`
+            The transition matrix
+        """
+        return self._transition_model(state).matrix(**kwargs)
+
+    def _smooth_gain(self, state, prediction, **kwargs):
+        """Calculate the smoothing gain
+
+        Parameters
+        ----------
+        state : :class:`~.State`
+            The input state
+        prediction : :class:`~.GaussianStatePrediction`
+            The prediction (from the subsequent state)
+
+        Returns
+        -------
+         : Matrix
+            The smoothing gain
+
+        """
+        return state.covar @ self._transition_matrix(state, **kwargs).T @ np.linalg.inv(
+            prediction.covar)
+
+    def smooth(self, track):
+        """
+        Perform the backward recursion to smooth the track.
+
+        Parameters
+        ----------
+        track : :class:`~.Track`
+            The input track.
+
+        Returns
+        -------
+         : :class:`~.Track`
+            Smoothed track
+
+        """
+        firststate = True
+        smoothed_track = Track()
+        for state in reversed(track):
+
+            if firststate:
+                subsq_state = state
+                smoothed_track.append(state)
+                firststate = False
+            else:
+                # Delta t
+                time_interval = subsq_state.timestamp - state.timestamp
+
+                # Retrieve the prediction from the subsequent (k+1th) timestep accessed previously
+                prediction = self._prediction(subsq_state)
+                # The smoothing gain, mean and covariance
+                ksmooth_gain = self._smooth_gain(state, prediction, time_interval=time_interval)
+                smooth_mean = state.state_vector + ksmooth_gain @ (subsq_state.state_vector -
+                                                                   prediction.state_vector)
+                smooth_covar = state.covar + \
+                    ksmooth_gain @ (subsq_state.covar - prediction.covar) @ ksmooth_gain.T
+
+                # Create a new type called SmoothedState?
+                if isinstance(state, Update):
+                    subsq_state = Update.from_state(state, smooth_mean, smooth_covar,
+                                                    timestamp=state.timestamp,
+                                                    hypothesis=state.hypothesis)
+                elif isinstance(state, Prediction):
+                    subsq_state = Prediction.from_state(state, smooth_mean, smooth_covar,
+                                                        timestamp=state.timestamp)
+
+                smoothed_track.append(subsq_state)
+
+        smoothed_track.reverse()
+        return smoothed_track
```

Comment:
Couple of suggestions below:
-  Take the first state bit out of the loop, removing need for `firststate`. (Change below looks bigger than it is, just to removing `if`/`else`)
- Create track at the end, using deep copy (need to import `copy`) to ensure meta data, id, etc. are kept; but using `memo` dictionary to avoid deep copying the original states (this would be very very expensive)

```suggestion
        subsq_state = track[-1]
        smoothed_states = [subsq_state]
        for state in reversed(track)[1:]:
            # Delta t
            time_interval = subsq_state.timestamp - state.timestamp

            # Retrieve the prediction from the subsequent (k+1th) timestep accessed previously
            prediction = self._prediction(subsq_state)
            # The smoothing gain, mean and covariance
            ksmooth_gain = self._smooth_gain(state, prediction, time_interval=time_interval)
            smooth_mean = state.state_vector + ksmooth_gain @ (subsq_state.state_vector -
                                                                   prediction.state_vector)
            smooth_covar = state.covar + \
                ksmooth_gain @ (subsq_state.covar - prediction.covar) @ ksmooth_gain.T

            # Create a new type called SmoothedState?
            if isinstance(state, Update):
                subsq_state = Update.from_state(state, smooth_mean, smooth_covar,
                                                timestamp=state.timestamp,
                                                hypothesis=state.hypothesis)
            elif isinstance(state, Prediction):
                subsq_state = Prediction.from_state(state, smooth_mean, smooth_covar,
                                                    timestamp=state.timestamp)

            smoothed_states.insert(0, subsq_state)

        # Deep copy existing track, but avoid copying original states, as this would be super
        # expensive. This works by informing deepcopy that the smoothed states are the
        # replacement object for the original track states.
        smoothed_track = copy.deepcopy(track, {id(track.states): smoothed_states})
        return smoothed_track
```

Reply:
`>       for state in reversed(track)[1:]:`
`E       TypeError: 'generator' object is not subscriptable`


`for state in reversed(track[:-1]):` seems to work ok though.